### ISSUE TITLE
dune cache writes merlin files to source tree with wrong umask

### ISSUE THREAD

**Issue by avsm** [ocaml/dune#3026]
With dune 2.1.1 and `DUNE_CACHE=enabled`, the `.merlin` files written to the source tree end up with the following permissions:

```
 % ls -la cli/.merlin 
-rw-r--r--  1 avsm  staff  3424 12 Jan 19:49 cli/.merlin
```

This subsequently results in dune segfaulting on a second build, since it fails the write to the .merlin file as it doesn't have write permissions for the user:

```
Error: exception Sys_error("cli/.merlin: Permission denied")
Backtrace:
Raised by primitive operation at file "stdlib.ml", line 324, characters 29-55
Called from file "src/stdune/io.ml", line 184, characters 8-126
Re-raised at file "src/stdune/io.ml", line 189, characters 8-23
Called from file "src/dune/artifact_substitution.ml", line 383, characters 15-48
Called from file "src/fiber/fiber.ml", line 229, characters 16-19
Called from file "src/dune/build_system.ml", line 1637, characters 8-1023
Called from file "src/fiber/fiber.ml", line 114, characters 10-15
Re-raised at file "src/stdune/exn.ml" (inlined), line 37, characters 27-56
Called from file "src/stdune/exn_with_backtrace.ml", line 13, characters 33-71
Called from file "src/fiber/fiber.ml", line 85, characters 10-17
Re-raised at file "src/stdune/exn.ml" (inlined), line 37, characters 27-56
Called from file "src/stdune/exn_with_backtrace.ml", line 13, characters 33-71
Called from file "src/fiber/fiber.ml", line 85, characters 10-17
Re-raised at file "src/stdune/exn.ml" (inlined), line 37, characters 27-56
Called from file "src/stdune/exn_with_backtrace.ml", line 13, characters 33-71
Called from file "src/fiber/fiber.ml", line 85, characters 10-17
Re-raised at file "src/stdune/exn.ml" (inlined), line 37, characters 27-56
Called from file "src/stdune/exn_with_backtrace.ml", line 13, characters 33-71
Called from file "src/fiber/fiber.ml", line 85, characters 10-17
Re-raised at file "src/stdune/exn.ml" (inlined), line 37, characters 27-56
Called from file "src/stdune/exn_with_backtrace.ml", line 13, characters 33-71
Called from file "src/fiber/fiber.ml", line 85, characters 10-17
Re-raised at file "src/stdune/exn.ml" (inlined), line 37, characters 27-56
Called from file "src/stdune/exn_with_backtrace.ml", line 13, characters 33-71
Called from file "src/fiber/fiber.ml", line 85, characters 10-17
Re-raised at file "src/stdune/exn.ml" (inlined), line 37, characters 27-56
Called from file "src/stdune/exn_with_backtrace.ml", line 13, characters 33-71
Called from file "src/fiber/fiber.ml", line 85, characters 10-17
Re-raised at file "src/stdune/exn.ml" (inlined), line 37, characters 27-56
Called from file "src/stdune/exn_with_backtrace.ml", line 13, characters 33-71
Called from file "src/fiber/fiber.ml", line 85, characters 10-17
Re-raised at file "src/stdune/exn.ml" (inlined), line 37, characters 27-56
Called from file "src/stdune/exn_with_backtrace.ml", line 13, characters 33-71
Called from file "src/fiber/fiber.ml", line 85, characters 10-17
Re-raised at file "src/stdune/exn.ml" (inlined), line 37, characters 27-56
Called from file "src/stdune/exn_with_backtrace.ml", line 13, characters 33-71
Called from file "src/fiber/fiber.ml", line 85, characters 10-17
Re-raised at file "src/stdune/exn.ml" (inlined), line 37, characters 27-56
Called from file "src/stdune/exn_with_backtrace.ml", line 13, characters 33-71
Called from file "src/fiber/fiber.ml", line 85, characters 10-17
Re-raised at file "src/stdune/exn.ml" (inlined), line 37, characters 27-56
Called from file "src/stdune/exn_with_backtrace.ml", line 13, characters 33-71
Called from file "src/fiber/fiber.ml", line 85, characters 10-17
```

This doesn't happen with DUNE_CACHE disabled as the merlin files have a more permissive umask.  This may be because the daemonisation of the dune daemon called `(Unix.umask 0)` so that the `~/.cache/dune` files are read-only, but this also ends up propagating to the `.merlin` files promoted to the source tree and causes a difference in behaviour with the dune cache disabled.

**Comment by diml** [comment]
That seems like a bug indeed. However:

> ```
>  % ls -la cli/.merlin 
> -rw-r--r--  1 avsm  staff  3424 12 Jan 19:49 cli/.merlin
> ```

I don't understand why this is not `-r--r--r--`, i.e. read-only for everyone. If it's writable by the user, why does dune fail to delete the file?

**Comment by avsm** [comment]
Sorry, I think I had a cut and paste on the bug report from when I confirmed it was working. It was read-only for the user as well with the dune cache enabled.

**Comment by diml** [comment]
Fixed in #3043

**Comment by mefyl** [comment]
As mentionned in https://github.com/ocaml/dune/pull/3043, we took the explicit decision to manually remove write permissions on all cached files, the rationale being:

* The cache can share files between build dirs with hardlinks, so editing a file would alter other build dirs which sounds horribly wrong.
* Editing produced files sounds like bad practice, as the build should be a computational graph.

I still think that decision is correct, and for reproducibility matter we should not edit produced files. Maybe we should tag such rules in dune as "creating a template file destined to be edited", which would make the targets non memoizable ?
